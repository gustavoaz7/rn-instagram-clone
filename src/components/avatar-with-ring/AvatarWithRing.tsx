import React, { memo } from 'react';
import { LinearGradient } from 'expo-linear-gradient';
import styled, { css, useTheme } from 'styled-components/native';

export type TAvatarWithRingProps = {
  size: number;
  color: string;
  imageUrl: string;
  ringWidth?: number;
  offset?: number;
};

export const AvatarWithRing = memo(function AvatarWithRing({
  size,
  color,
  imageUrl,
  ringWidth = 2,
  offset = 2,
}: TAvatarWithRingProps) {
  const theme = useTheme();
  const gradientColors =
    color === 'gradient'
      ? [theme.color.purpleRed, theme.color.yellow]
      : [color, color];

  return (
    <Container size={size}>
      <GradientRing
        colors={gradientColors}
        start={{ x: 0.7, y: 0.3 }}
        end={{ x: 0.3, y: 0.7 }}
      >
        <AvatarContainer size={size} ringWidth={ringWidth}>
          <Avatar
            size={size}
            ringWidth={ringWidth}
            offset={offset}
            source={{ uri: imageUrl }}
          />
        </AvatarContainer>
      </GradientRing>
    </Container>
  );
});

const centerStyle = css`
  align-items: center;
  justify-content: center;
`;

const Container = styled.View<Pick<TAvatarWithRingProps, 'size'>>`
  position: relative;
  overflow: hidden;
  ${({ size }) => `
    height: ${size}px;
    width: ${size}px;
    border-radius: ${size}px;
  `}
`;

const GradientRing = styled(LinearGradient)`
  width: 100%;
  height: 100%;
  ${centerStyle};
`;

const AvatarContainer = styled.View<Record<'size' | 'ringWidth', number>>`
  ${centerStyle};
  background-color: ${({ theme }) => theme.color.white};
  ${({ size, ringWidth }) => `
    height: ${size - ringWidth * 2}px;
    width: ${size - ringWidth * 2}px;
    border-radius: ${size - ringWidth * 2}px;
  `}
`;

const Avatar = styled.Image<Record<'size' | 'ringWidth' | 'offset', number>>`
  ${({ size, ringWidth, offset, theme }) => `
    height: ${size - ringWidth * 2 - offset * 2}px;
    width: ${size - ringWidth * 2 - offset * 2}px;
    border-radius: ${size - ringWidth * 2 - offset * 2}px;
    border-width: 2px;
    border-color: ${`${theme.color.black}33`};
  `}
`;
