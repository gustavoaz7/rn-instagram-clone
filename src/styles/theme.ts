import type { Animated } from 'react-native';

const HEART_SPRING_SCALE_CONFIG: Animated.SpringAnimationConfig = {
  toValue: 1,
  useNativeDriver: true,
  velocity: 5,
  tension: 100,
  friction: 5,
};

export const theme = {
  spacing: {
    xs: '5px',
    s: '10px',
    m: '15px',
    l: '20px',
    xl: '25px',
  },
  color: {
    lightBlue: '#0095f6',
    royalBlue: '#405DE6',
    blue: '#5B51D8',
    purple: '#833AB4',
    darkPink: '#C13584',
    purpleRed: '#E1306C',
    red: '#FD1D1D',
    darkOrange: '#F56040',
    orange: '#F77737',
    yellow: '#FCAF45',
    lightYellow: '#FFDC80',
    white: '#FFFFFF',
    black: '#262626',
    gray: '#8e8e8e',
  },
  font: {
    size: {
      s: '12px',
      m: '14px',
      l: '16px',
      xl: '20px',
    },
  },
  animation: {
    timingFast: 150,
    timingBase: 300,
    timingSlow: 500,

    heart: {
      initialScale: 0.3,
      springConfig: HEART_SPRING_SCALE_CONFIG,
    },
  },
};

export type TTheme = typeof theme;
