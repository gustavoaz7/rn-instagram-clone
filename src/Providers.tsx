import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import { theme as stylesTheme } from './styles/theme';
import { store as reduxStore } from './redux/store';

type TProvidersProps = {
  children: ReactNode;
  store?: typeof reduxStore;
  theme?: typeof stylesTheme;
};

export function Providers({
  children,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>
  );
}
