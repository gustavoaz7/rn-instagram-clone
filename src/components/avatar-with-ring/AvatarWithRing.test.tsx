import React from 'react';
import { render } from '@testing-library/react-native';
import { LinearGradient } from 'expo-linear-gradient';
import { AvatarWithRing } from './AvatarWithRing';
import { Providers } from '../../Providers';
import { theme } from '../../styles/theme';

describe('components - AvatarWithRing', () => {
  const options = { wrapper: Providers };
  const props = {
    size: 70,
    color: 'black',
    imageUrl: 'https://cdn.fakercloud.com/avatars/spbroma_128.jpg',
  };

  it('renders', () => {
    render(<AvatarWithRing {...props} />, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<AvatarWithRing {...props} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot for custom ringWidth and offset', () => {
    const { toJSON } = render(
      <AvatarWithRing {...props} offset={5} ringWidth={1} />,
      options,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders ring with provided color', () => {
    const color = 'green';
    const { UNSAFE_getByType } = render(
      <AvatarWithRing {...props} color={color} />,
      options,
    );

    expect(UNSAFE_getByType(LinearGradient).props.colors).toEqual([
      color,
      color,
    ]);
  });

  it('renders gradient ring when "gradient" color is provided', () => {
    const { UNSAFE_getByType } = render(
      <AvatarWithRing {...props} color="gradient" />,
      options,
    );

    expect(UNSAFE_getByType(LinearGradient).props.colors).toEqual([
      theme.color.purpleRed,
      theme.color.yellow,
    ]);
  });
});
