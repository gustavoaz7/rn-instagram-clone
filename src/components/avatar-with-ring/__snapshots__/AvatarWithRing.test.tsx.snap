// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - AvatarWithRing matches snapshot 1`] = `
<View
  size={70}
  style={
    Array [
      Object {
        "borderRadius": 70,
        "height": 70,
        "overflow": "hidden",
        "position": "relative",
        "width": 70,
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4278190080,
          4278190080,
        ],
        "endPoint": Array [
          0.3,
          0.7,
        ],
        "locations": undefined,
        "startPoint": Array [
          0.7,
          0.3,
        ],
      }
    }
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <View
      ringWidth={2}
      size={70}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 66,
            "height": 66,
            "justifyContent": "center",
            "width": 66,
          },
        ]
      }
    >
      <Image
        offset={2}
        ringWidth={2}
        size={70}
        source={
          Object {
            "uri": "https://cdn.fakercloud.com/avatars/spbroma_128.jpg",
          }
        }
        style={
          Array [
            Object {
              "borderColor": "#26262633",
              "borderRadius": 62,
              "borderWidth": 2,
              "height": 62,
              "width": 62,
            },
          ]
        }
      />
    </View>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;

exports[`components - AvatarWithRing matches snapshot for custom ringWidth and offset 1`] = `
<View
  size={70}
  style={
    Array [
      Object {
        "borderRadius": 70,
        "height": 70,
        "overflow": "hidden",
        "position": "relative",
        "width": 70,
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4278190080,
          4278190080,
        ],
        "endPoint": Array [
          0.3,
          0.7,
        ],
        "locations": undefined,
        "startPoint": Array [
          0.7,
          0.3,
        ],
      }
    }
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <View
      ringWidth={1}
      size={70}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 68,
            "height": 68,
            "justifyContent": "center",
            "width": 68,
          },
        ]
      }
    >
      <Image
        offset={5}
        ringWidth={1}
        size={70}
        source={
          Object {
            "uri": "https://cdn.fakercloud.com/avatars/spbroma_128.jpg",
          }
        }
        style={
          Array [
            Object {
              "borderColor": "#26262633",
              "borderRadius": 58,
              "borderWidth": 2,
              "height": 58,
              "width": 58,
            },
          ]
        }
      />
    </View>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
